### PR TITLE
KAA-1571: [C] Introduce CMake toolchain file for Intel Edison platform.

### PR DESCRIPTION
**Pull request by RostakaGmfun**
This commit also slightly corrects this toolchain file.

/cc @Slash32 @vadimol

### CODE REVIEW

File: client/client-multi/client-cpp/toolchains/edison.cmake [hunk 1]
```diff
@@ -23,19 +23,23 @@ set(CMAKE_SYSTEM_NAME Linux)
...
-	else()
-		set(EDISON_SDK_ROOT C:/Edison/poky-edison-eglibc-x86_64-edison-image-core2-32-toolchain-1.7.2)
-	endif()
-else(NOT DEFINED EDISON_SDK_ROOT)
-	set(EDISON_SDK_ROOT /opt/poky-edison/1.6)
-endif()
+if(NOT DEFINED EDISON_SDK_ROOT)
+    if(WIN32)
+        if(${SDK_32bits})
+            set(EDISON_SDK_ROOT C:/Edison/poky-edison-eglibc-i686-edison-image-core2-32-toolchain-1.7.2)
```
Comment (Slash32) [inline client/client-multi/client-cpp/toolchains/edison.cmake]: This should be better done with use of Environment variables to allow users to change the place.

File: client/client-multi/client-cpp/toolchains/edison.cmake [hunk 2]
```diff
@@ -23,19 +23,23 @@ set(CMAKE_SYSTEM_NAME Linux)
...
+        else()
+            set(EDISON_SDK_ROOT C:/Edison/poky-edison-eglibc-x86_64-edison-image-core2-32-toolchain-1.7.2)
```
Comment (Slash32) [inline client/client-multi/client-cpp/toolchains/edison.cmake]: Same as above

File: client/client-multi/client-cpp/toolchains/edison.cmake [hunk 3]
```diff
@@ -23,19 +23,23 @@ set(CMAKE_SYSTEM_NAME Linux)
...
+        endif()
+    else(WIN32)
+        set(EDISON_SDK_ROOT /opt/poky-edison/1.6)
```
Comment (Slash32) [inline client/client-multi/client-cpp/toolchains/edison.cmake:50]: Same as above

File: client/client-multi/client-cpp/toolchains/edison.cmake [hunk 4]
```diff
@@ -46,7 +50,7 @@ if(WIN32)
 		set(CMAKE_MAKE_PROGRAM "C:/Program Files (x86)/mingw/i686-4.9.2-posix-dwarf-rt_v3-rev1/mingw32/bin/mingw32-make.exe")
```
Comment (Slash32) [inline client/client-multi/client-cpp/toolchains/edison.cmake]: Same as above

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    else(WIN32)
+        set(EDISON_SDK_ROOT /opt/poky-edison/1.6)
+    endif(WIN32)
+endif(NOT DEFINED EDISON_SDK_ROOT)
+
+message(STATUS "EDISON_SDK_ROOT: ${EDISON_SDK_ROOT}")
+
+#Set the host parameters
+if(WIN32)
+    #Windows host
```
Comment (Slash32) [inline client/client-multi/client-c/toolchains/edison.cmake]: Indentation issue

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	if(${SDK_32bits})
+		set(cross_compiler_sysroot ${EDISON_SDK_ROOT}/sysroots/i686-pokysdk-mingw32)
+		#MinGW make
+		set(CMAKE_MAKE_PROGRAM "C:/Program Files/mingw/i686-4.9.2-posix-dwarf-rt_v3-rev1/mingw32/bin/mingw32-make.exe")
```
Comment (Slash32) [inline client/client-multi/client-c/toolchains/edison.cmake]: Please use Environment variables for this (see comments below).

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	else()
+		set(cross_compiler_sysroot ${EDISON_SDK_ROOT}/sysroots/x86_64-pokysdk-mingw32)
+		#MinGW make
+		set(CMAKE_MAKE_PROGRAM "C:/Program Files (x86)/mingw/i686-4.9.2-posix-dwarf-rt_v3-rev1/mingw32/bin/mingw32-make.exe")
```
Comment (Slash32) [inline client/client-multi/client-c/toolchains/edison.cmake]: Same as above

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+# limitations under the License.
+#
+
+if(NOT DEFINED SDK_32bits)
+	set(SDK_32bits FALSE)
+endif()
+
+#Set the target parameters
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSTEM_VERSION 3.10.17-poky-edison+)
```
Comment (Slash32) [inline client/client-multi/client-c/toolchains/edison.cmake:31]: Please ensure this version is also stated in the documentation.

**Review by Slash32 — CHANGES_REQUESTED** [review]

File: client/client-multi/client-c/toolchains/edison.cmake (hunk 5 above)
Comment (RostakaGmfun) [reply client/client-multi/client-c/toolchains/edison.cmake]: Don't see anything wrong

File: client/client-multi/client-cpp/toolchains/edison.cmake (hunk 2 above)
Comment (RostakaGmfun) [reply client/client-multi/client-cpp/toolchains/edison.cmake]: Done

File: client/client-multi/client-cpp/toolchains/edison.cmake (hunk 3 above)
Comment (RostakaGmfun) [reply client/client-multi/client-cpp/toolchains/edison.cmake:50]: Done

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        set(EDISON_SDK_ROOT /opt/poky-edison/1.6)
+    endif(WIN32)
+else(NOT EXISTS "$ENV{EDISON_SDK_ROOT}")
+    set(EDISON_SDK_ROOT $ENV{EDISON_SDK_ROOT})
+endif(NOT EXISTS "$ENV{EDISON_SDK_ROOT}")
+
+message(STATUS "EDISON_SDK_ROOT: ${EDISON_SDK_ROOT}")
+
+#Set the host parameters
+if(WIN32)
```
Comment (vadimol) [inline client/client-multi/client-c/toolchains/edison.cmake]: Have you tested this on Windows platform?

**Review by vadimol — CHANGES_REQUESTED** [review]

File: client/client-multi/client-c/toolchains/edison.cmake (hunk 8 above)
Comment (RostakaGmfun) [reply client/client-multi/client-c/toolchains/edison.cmake:31]: Done

File: client/client-multi/client-c/toolchains/edison.cmake (hunk 9 above)
Comment (RostakaGmfun) [reply client/client-multi/client-c/toolchains/edison.cmake]: Nope.

Comment (vadimol) [reply client/client-multi/client-c/toolchains/edison.cmake]: Then remove all the code that relates to Windows support. Let's avoid committing the code that does not work.

**Review by vadimol — CHANGES_REQUESTED** [review]
See comment

File: client/client-multi/client-c/toolchains/edison.cmake (hunk 9 above)
Comment (RostakaGmfun) [reply client/client-multi/client-c/toolchains/edison.cmake]: Great idea!

File: doc/Programming-guide/Using-Kaa-endpoint-SDKs/C/SDK-Edison/index.md [hunk 10]
```diff
@@ -44,8 +44,7 @@ Now, dependencies are installed and it is time to create Kaa application.
 Since Edison is running Linux, you can refer to [the Linux guide]({{root_url}}Programming-guide/Using-Kaa-endpoint-SDKs/C/SDK-Linux/#c-sdk-build) for detailed process of application creation.
 But remember, you must specify correct compiler when compiling your Kaa application for Intel Edison:
 
-        export EDISON_CC=/opt/poky-edison/1.6.1/sysroots/x86_64-pokysdk-linux/usr/bin/i586-poky-linux/i586-poky-linux-gcc
-        cmake -DKAA_MAX_LOG_LEVEL=3 -DCMAKE_C_COMPILER=$EDISON_CC ..
+        cmake -DKAA_MAX_LOG_LEVEL=3 -DCMAKE_TOOLCHAIN_FILE=../libs/kaa/toolchains/edison.cmake ..
```
Comment (vadimol) [inline doc/Programming-guide/Using-Kaa-endpoint-SDKs/C/SDK-Edison/index.md]: Why libs/kaa? We should not assume that user compiles our sample application.

**Review by vadimol — CHANGES_REQUESTED** [review]
One minor nitpick. Overall looks good.

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+if(NOT DEFINED SDK_32bits)
+	set(SDK_32bits FALSE)
```
Comment (rasendubi) [inline client/client-multi/client-c/toolchains/edison.cmake]: Use spaces, not tabs.

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+endif()
+
+#Set the target parameters
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSTEM_VERSION 3.10.17-poky-edison+)
+set(CMAKE_SYSTEM_PROCESSOR i686)
+
+if(NOT EXISTS "$ENV{EDISON_SDK_ROOT}")
```
Comment (rasendubi) [inline client/client-multi/client-c/toolchains/edison.cmake]: We prefer CMake variable over environment variables.

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        set(EDISON_SDK_ROOT /opt/poky-edison/1.6)
+else(NOT EXISTS "$ENV{EDISON_SDK_ROOT}")
```
Comment (rasendubi) [inline client/client-multi/client-c/toolchains/edison.cmake]: Don't put condition in `else()` and `endif()` macros.

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+
+set(CMAKE_SYSROOT ${EDISON_SDK_ROOT}/sysroots/core2-32-poky-linux)
+set(CMAKE_FIND_ROOT_PATH ${CMAKE_SYSROOT})
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
+
+SET(CMAKE_C_FLAGS "-Os -g3 -Wall --sysroot=${CMAKE_SYSROOT} -m32 -march=i586 -ffunction-sections -fdata-sections"  CACHE STRING "" FORCE)
```
Comment (rasendubi) [inline client/client-multi/client-c/toolchains/edison.cmake]: `-Os -g3 -Wall` do not belong here.

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+SET(CMAKE_CXX_FLAGS "-Os -g3 -Wall --sysroot=${CMAKE_SYSROOT} -m32 -march=i586 -ffunction-sections -fdata-sections"  CACHE STRING "" FORCE)
```
Comment (rasendubi) [inline client/client-multi/client-c/toolchains/edison.cmake]: `-Os -g3 -Wall` do not belong here.

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+INCLUDE_DIRECTORIES(${CMAKE_SYSROOT}/usr/include)
```
Comment (rasendubi) [inline client/client-multi/client-c/toolchains/edison.cmake]: Use lower-case for macro names.

**Review by rasendubi — COMMENTED** [review]
Check code style. Also, why toolchains for C and C++ SDKs are different?

I wonder if we can merge them into a single file. (I know that's hard now, but we should think about that in the future.)

File: client/client-multi/client-c/toolchains/edison.cmake [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+set(CMAKE_SYSTEM_VERSION 3.10.17-poky-edison+)
+set(CMAKE_SYSTEM_PROCESSOR i686)
+
+if(NOT DEFINED EDISON_SDK_ROOT)
+    set(EDISON_SDK_ROOT /opt/poky-edison/1.6)
+endif()
+
+message(STATUS "EDISON_SDK_ROOT: ${EDISON_SDK_ROOT}")
+
+if(${SDK_32bits})
```
Comment (rasendubi) [inline client/client-multi/client-c/toolchains/edison.cmake]: Avoid double-expansion.

``` cmake
if(SDK_32bits)
```

File: client/client-multi/client-cpp/toolchains/edison.cmake [hunk 18]
```diff
@@ -13,40 +13,65 @@
...
+        if(${SDK_32bits})
+            set(CMAKE_MAKE_PROGRAM "C:/Program Files/mingw/i686-4.9.2-posix-dwarf-rt_v3-rev1/mingw32/bin/mingw32-make.exe")
+        else(${SDK_32bits})
+            set(CMAKE_MAKE_PROGRAM "C:/Program Files (x86)/mingw/i686-4.9.2-posix-dwarf-rt_v3-rev1/mingw32/bin/mingw32-make.exe")
+        endif(${SDK_32bits})
+    else()
+        set(CMAKE_MAKE_PROGRAM "${EDISON_MAKE_PROGRAM}")
+    endif()
+    #Windows host
+    if(${SDK_32bits})
```
Comment (rasendubi) [inline client/client-multi/client-cpp/toolchains/edison.cmake]: See comment above.

**Review by rasendubi — COMMENTED** [review]
Otherwise, looks good.

**Review by rasendubi — APPROVED** [review]

**Review by vadimol — APPROVED** [review]

**Pull request merged**